source/common/api/posix/os_sys_calls_impl.cc

```diff
@@ -73,6 +73,18 @@ bool OsSysCallsImpl::supportsMmsg() const {
 #endif
 }
 
+bool OsSysCallsImpl::supportsUdpGro() const {
+#if !defined(__linux__)
+  return false;
+#else
+#ifndef UDP_GRO
+  return false;
+#else
+  return true;
+#endif
```

Comment:
I think this is not enough, it's too easy to define UDP_GRO somewhere without kernel support. Also a binary built with UDP_GRO will not work correctly on old kernels. You can create a udp socket here and try to setsocketopt to test if it's supported, you can use a static variable to only test it once:

```
static const bool is_supported = [] {
  int fd = socket(AF_INET, SOCK_DGRAM | SOCK_NONBLOCK, IPPROTO_UDP);
  if (fd < 0) {
    return false;
  }
  bool result = (0 == setsockopt(fd, IPPROTO_UDP, UDP_GRO, &val, sizeof(val)));
  close(fd);
  return result;
}();
return is_supported;
```

Reply:
Thanks for pointing this out! I have now added the setsocketopt test instead of simply returning true. 
```
bool OsSysCallsImpl::supportsUdpGro() const {
#if !defined(__linux__)
  return false;
#else
#ifndef UDP_GRO
  return false;
#else
  static const bool is_supported = [] {
    int fd = ::socket(AF_INET, SOCK_DGRAM | SOCK_NONBLOCK, IPPROTO_UDP);
    if (fd < 0) {
      return false;
    }
    int val = 1;
    bool result = (0 == ::setsockopt(fd, IPPROTO_UDP, UDP_GRO, &val, sizeof(val)));
    ::close(fd);
    return result;
  }();
  return is_supported;
#endif 
#endif
}
```

Reply:
Continuing on the discussion that we had regarding checking kernel versions appropriately, I performed all three tests and built the test on both RBE(v 4.18) and cloudtop(v 5.2). Kernel Version tests used were:

1. linux/version.h.               [ Compile Time Linux Version]
2. Using sys/utsname.h      [ Run Time Linux Version ]
3. setsockopt                      [ Run Time setsockopt test ]

The following are the logs displaying the results of the test run on RBE:
```
Yugant: LINUX version is : 4                  <=== Check using linux/version.h                                                                                                                                        
Yugant: UTS version is: 5                     <=== Check using sys/utsname.h                                                                                                                                
Yugant: setsockopt supported here             <=== Check using setsockopt
```

Similarly after compiling on cloudtop, the interesting logs are as follows:
```
Yugant: LINUX version is : 5                                                                                                                                              
Yugant: UTS version is: 5                                                                                                                                                    
Yugant: Supported here 
```

I looks like for remote RBE execution, bazel compiles the code on the RBE, (Hence, the Linux version shows up as 4). But the compiled test under both the cases (RBE compile and cloudtop compile) is run locally on the cloudtop (hence uts version shows as 5).

Based on this, I think we can safely use either of setsockopt test or UTS version test for under supportsUdpGro. (Currently I am going ahead with setsockopt test)

Please let me know if you have some other thoughts regarding this.

Reply:
Based on the recent discussions, going ahead with the use of setsockopt(...) test here!